src/blockchain/alt_block_tree.cpp

```diff
@@ -460,37 +463,47 @@ bool AltTree::addPayloads(AltTree::index_t& index,
   return addPayloads(index, copy, state, false);
 }
 
-bool AltTree::setTip(AltTree::index_t& to,
-                     ValidationState& state,
-                     bool skipSetState,
-                     bool continueOnInvalid) {
-  bool changeTip = true;
-  if (!skipSetState) {
-    changeTip = cmp_.setState(*this, to, state, continueOnInvalid);
-    if (continueOnInvalid) {
-      VBK_ASSERT(changeTip);
-    }
+bool AltTree::setState(index_t& to, ValidationState& state) {
+  bool success = cmp_.setState(*this, to, state);
+  if (success) {
+    bool success = overrideTip(to, state);
+    VBK_ASSERT(success);
+  } else {
+    // if setState failed, then 'to' must be invalid
+    VBK_ASSERT(!to.isValid());
   }
+  return success;
+}
+
+bool AltTree::overrideTip(index_t& to, ValidationState& state) {
+  VBK_LOG_INFO("ALT=\"%s\", VBK=\"%s\", BTC=\"%s\"",
+               to.toShortPrettyString(),
+               (vbk().getBestChain().tip()
+                    ? vbk().getBestChain().tip()->toShortPrettyString()
+                    : "<empty>"),
+               (btc().getBestChain().tip()
+                    ? btc().getBestChain().tip()->toShortPrettyString()
+                    : "<empty>"));
+  activeChain_.setTip(&to);
+  tryAddTip(&to);
+  return true;
+}
+
+bool AltTree::setTipContinueOnInvalid(AltTree::index_t& to,
+                                      ValidationState& state) {
+  bool success = cmp_.setState(*this, to, state, /*continueOnInvalid=*/true);
+  VBK_ASSERT(success);
 
-  // edge case: if changeTip is false, then new block arrived on top of
-  // current active chain, and this block has invalid commands
-  if (changeTip) {
-    VBK_LOG_INFO("ALT=\"%s\", VBK=\"%s\", BTC=\"%s\"",
-                 to.toShortPrettyString(),
-                 (vbk().getBestChain().tip()
-                      ? vbk().getBestChain().tip()->toShortPrettyString()
-                      : "<empty>"),
-                 (btc().getBestChain().tip()
-                      ? btc().getBestChain().tip()->toShortPrettyString()
-                      : "<empty>"));
-    activeChain_.setTip(&to);
-    tryAddTip(&to);
+  if (success) {
+    bool success = overrideTip(to, state);
+    VBK_ASSERT(success);
   } else {
+    // if setState failed, then 'to' must be invalid
     VBK_ASSERT(!to.isValid());
   }
```

Comment:
no need in this condition, because `success` is always true

Reply:
setState is definitely not always going to succeed

Reply:
look closely - this is setTipContinueOnInvalid, it always succeeds, and there are two asserts that indicate this
https://github.com/VeriBlock/alt-integration-cpp/pull/291/files#diff-2cf3b8cf328ee87b5b03c1b7d0abf949R481
https://github.com/VeriBlock/alt-integration-cpp/pull/291/files#diff-2cf3b8cf328ee87b5b03c1b7d0abf949R485

Reply:
ah right. github doesn't show enough context lines :(